src/Pact/Types/ExpParser.hs

```diff
@@ -0,0 +1,263 @@
+  takeWhile_ f Cursor{..} = second (Cursor _cContext) $ span f _cStream
+
+
+data ParseState a = ParseState
+  { _psCurrent :: (Exp Info)
+  , _psUser :: a
+  }
+makeLenses ''ParseState
+
+
+type MkInfo = Parsed -> Info
+
+{-# INLINE mkEmptyInfo #-}
+mkEmptyInfo :: MkInfo
+mkEmptyInfo e = Info (Just (mempty,e))
+
+{-# INLINE mkStringInfo #-}
+mkStringInfo :: String -> MkInfo
+mkStringInfo s d = Info (Just (fromString $ take (_pLength d) $
+                               drop (fromIntegral $ TF.bytes d) s,d))
+
+{-# INLINE mkTextInfo #-}
+mkTextInfo :: T.Text -> MkInfo
+mkTextInfo s d = Info (Just (Code $ T.take (_pLength d) $
+                             T.drop (fromIntegral $ TF.bytes d) s,d))
+
+type ExpParse s a = StateT (ParseState s) (Parsec Void Cursor) a
+
+
+{-# INLINE runCompile #-}
+runCompile :: ExpParse s a -> ParseState s -> Exp Info -> Either PactError a
+runCompile act cs a =
+  case runParser (runStateT act cs) "" (Cursor Nothing [a]) of
+    (Right (r,_)) -> Right r
+    (Left (TrivialError _ (Just err) expect)) -> case err of
+      EndOfInput -> case S.toList expect of
+        (Tokens (x :| _):_) -> doErr (eInfo x) "unexpected end of input"
+        (Label s:_) -> doErr def (toList s)
+        er -> doErr def (show er)
+      Label ne -> doErr def (toList ne)
+      Tokens (x :| _) -> doErr (eInfo x) $ showExpect expect
+    (Left e) -> doErr def (show e)
+    where doErr i s = Left $ PactError SyntaxError i def (pack s)
+          showExpect e = case (labelText $ S.toList e) of
+            [] -> show (S.toList e)
+            ss -> intercalate "," ss
+          labelText [] = []
+          labelText (Label s:r) = toList s:labelText r
+          labelText (EndOfInput:r) = "End of input":labelText r
+          labelText (_:r) = labelText r
+
+
+{-# INLINE strErr #-}
+strErr :: String -> ErrorItem t
+strErr = Label . fromList
+
+{-# INLINE tokenErr #-}
+tokenErr :: String -> Exp Info -> ExpParse s a
+tokenErr s = tokenErr' s . Just
+
+{-# INLINE tokenErr' #-}
+tokenErr' :: String -> Maybe (Exp Info) -> ExpParse s a
+tokenErr' ty i = failure
+  (fmap (\e -> (Tokens (e:|[]))) i)
+  (S.singleton (strErr ty))
+
+{-# INLINE context #-}
+context :: ExpParse s (Maybe (Exp Info))
+context = (fmap snd . _cContext) <$> getInput
+
+{-# INLINE contextInfo #-}
+contextInfo :: ExpParse s Info
+contextInfo = maybe def eInfo <$> context
+
+{-# INLINE current #-}
+current :: ExpParse s (Exp Info)
+current = use psCurrent
+
+{-# INLINE syntaxError #-}
+syntaxError :: String -> ExpParse s a
+syntaxError s = current >>= tokenErr s
+
+{-# INLINE expected #-}
+expected :: String -> ExpParse s a
+expected s = syntaxError $ "Expected: " ++ s
+
+{-# INLINE unexpected' #-}
+unexpected' :: String -> ExpParse s a
+unexpected' s = syntaxError $ "Unexpected: " ++ s
+
+
+{-# INLINE exp #-}
+exp :: String -> Prism' (Exp Info) a -> (a-> Either String b) -> ExpParse s (b,Exp Info)
+exp ty prism test2 = do
```

Comment:
_(Re-adding this comment, because the previous one got wiped out)_

Maybe rename `test2` something like `handle`?